src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -481,14 +482,40 @@ public ISet<string> WarningsAsMessages
         public bool HasBuildSubmissionLoggedErrors(int submissionId)
         {
             // Warnings as errors are not tracked if the user did not specify to do so
-            if (WarningsAsErrors == null)
+            if (WarningsAsErrors == null && _filterEventSource.WarningsAsErrorsByProject == null && _eventSinkDictionary.Values.All(i => i.WarningsAsErrorsByProject == null))
             {
                 return false;
             }
 
             // Determine if any of the event sinks have logged an error with this submission ID
-            return (_filterEventSource != null && _filterEventSource.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId))
-                || (_eventSinkDictionary != null && _eventSinkDictionary.Values.Any(i => i.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId)));
+            return _filterEventSource != null && _filterEventSource.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId)
+                || _eventSinkDictionary != null && _eventSinkDictionary.Values.Any(i => i.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId));
+        }
+
+        public void AddWarningsAsErrors(int projectInstanceId, ISet<string> codes)
+        {
+            lock (_lockObject)
+            {
+                if (_filterEventSource.WarningsAsErrorsByProject == null)
+                {
+                    _filterEventSource.WarningsAsErrorsByProject = new ConcurrentDictionary<int, ISet<string>>();
+                }
+
+                _filterEventSource.WarningsAsErrorsByProject.Add(projectInstanceId, new HashSet<string>(codes, StringComparer.OrdinalIgnoreCase));
+            }
+        }
+
+        public void AddWarningsAsMessages(int projectInstanceId, ISet<string> codes)
+        {
+            lock (_lockObject)
```

Comment:
If you are using a concurrent dictionary, rather than locking, you can use a concurrent Lazy to initialize it, and then not use the lock.